README.md

```diff
@@ -74,13 +74,16 @@ Note that depending on your syntax highlighter, you might have to include additi
 
 ## Props
 
+The editor accepts all the props accepted by `textarea`. In addition, you can pass the following props:
+
 - `value` (`string`): Current value of the editor i.e. the code to display. This must be a [controlled prop](https://reactjs.org/docs/forms.html#controlled-components).
 - `onValueChange` (`string => mixed`): Callback which is called when the value of the editor changes. You'll need to update the value prop when this is called.
 - `highlight` (`string => string | React.Node`): Callback which will receive text to highlight. You'll need to return an HTML string or a React element with syntax highlighting using a library such as [`prismjs`](https://prismjs.com).
 - `tabSize` (`number`): The number of characters to insert when pressing tab key. For example, for 4 space indentation, `tabSize` will be `4` and `insertSpaces` will be `true`. Default: `2`.
 - `insertSpaces` (`boolean`): Whether to use spaces for indentation. Default: `true`.
 - `ignoreTabKey` (`boolean`): Whether the editor should ignore tab key presses so that keyboard users can tab past the editor. Users can toggle this behaviour using `Ctrl+Shift+M` (Mac) / `Ctrl+M` manually when this is `false`. Default: `false`.
 - `padding` (`number`): Optional padding for code. Default: `0`.
+- `children` (`(props, styles) => React.Node`): Optional render callback to provide a custom implementation for the editor. Receives some props to pass to the `textarea` and styles to apply to the `textarea` and the `pre` element. See [`src/index.js`](src/index.js) for the default implementation.
```

Comment:
So I personally really _love love love_ prop getters. They make it explicit which props to pass where and how to override props without breaking anything.

`getTextarea`, `getPre`, and `getWrapper` sounds quite sensible. This has the advantage that when they receive `{ style: x }` you can decide how to treat `x` in the context of the styles that _must_ be applied.

Reply:
From chat:

>  they're really a great way to write render props since you can specifically differentiate between "user input" and your "prop output". It also becomes obvious which props go where as you'd have one prop getter per element
> Alternatively [or additonally,] it might be cool to reverse the pattern; simply have transform{Textarea,Pre,Wrapper}Props and then render out with three arguments with a default children function. [This is neat because the typical three elements don't change much as they're required for this library to work well. But that could be some sugar on top of the render prop for a separate PR].
> Either way prop getters really cut down on the amount of code the user has to write, because essentially everything that is now in your default children fn needs to be copied over even if just one thing is supposed to be modified [as opposed to getters accepting an object of "overrides" that can readily be combined in the case of `style` etc]

Reply:
Do you think functions are necessary for the props here? I could just pass the props directly, like `{...props.wrapper}` or `{...props.textarea}` and `{...props.pre}`. they look a bit nicer than `{...getWrapperProps()}` and I don't have a reason to make them functions.

> Alternatively [or additonally,] it might be cool to reverse the pattern; simply have transform{Textarea,Pre,Wrapper}Props and then render out with three arguments with a default children function

Can you expand a bit on this? I didn't understand what you meant.

Reply:
Having a function gives you a lot more flexibility to transform whatever comes in. It'd be possible to alternatively even accept your `padding` prop in the style functions which makes it super transparent where it actually takes effect.

Essentially I'd see this as an opt-in pattern for `react-simple-code-editor`, where the previous API is not "disturbed" but can be broken down into a render-prop API.

The prop getters make this part less painful as it yet again abstracts more of the "internals".

Reply:
As to the reversal:

```js
<Editor
  transformTextareaProps={props => ({ ...props, id: 'editor-textarea' })}
  transformPreProps={props => ({ ...props, className: 'test' })}
/>
```

See second props: It still could give you the opportunity to merge `className`. I'd just be a little disappointed if you'd offload this to the user when they decide to write their own `children` render prop function :)

Reply:
Hey folks, am hoping to nudge this along because I think it's a wonderful improvement... any possibility of approving this pull request, and then later it can be upgraded to use a propGetter situation?